### PR TITLE
SPLAT-1637: Initial changes for multi vcenter support.

### PR DESCRIPTION
**Pull request by vr4manta**
[SPLAT-1637](https://issues.redhat.com/browse/SPLAT-1637)

### Changes
- Bumped openshift/api to get new feature gate
- Stubbed in initial changes to allow multi vcenter to prevent operator from becoming degraded.
- Move user/pass from env vars to csi config file due to not allowing multiple user/pass as env vars (messy).
- Enhanced CSI Config file for multi vcenter config and password for each due to needing ability to configure each vCenter with unique user/pass.
- Updated checks to include a connection for all vCenters.
- Updated Unit Test

### Prerequisites 
- https://github.com/openshift/cluster-storage-operator/pull/465
- https://github.com/openshift/api/pull/1842
- https://github.com/openshift/vmware-vsphere-csi-driver-operator/pull/227
- https://github.com/openshift/vmware-vsphere-csi-driver/pull/122

### Blocks
- https://github.com/openshift/installer/pull/8638

### Notes
There is a PR changing the config to a secret which helps me out with this PR.  I am setting user/pass in the csi config just like it and was about to change it from configmap to secret.  The PR is: https://github.com/openshift/vmware-vsphere-csi-driver-operator/pull/227

### CODE REVIEW

**Comment by vr4manta** [comment]
/assign @gnufied 
Ready for review.

File: pkg/operator/vspherecontroller/vspherecontroller.go [hunk 1]
```diff
@@ -456,32 +465,44 @@ func (c *VSphereController) createVCenterConnection(ctx context.Context, infra *
...
+		if !ok {
+			return fmt.Errorf("error parsing secret %q: key %q not found", cloudCredSecretName, passwordKey)
+		}
+
+		// just a hack for other function compatibility.
+		/*cfg := new(legacy.VSphereConfig)
+		cfg.Workspace.VCenterIP = vcenter.Server
+		cfg.Workspace.Datacenter = vcenter.Datacenters[0]
+		cfg.Workspace.DefaultDatastore = infra.Spec.PlatformSpec.VSphere.FailureDomains[0].Topology.Datastore
+		cfg.Global.InsecureFlag = true*/
```
Comment (rvanderp3) [inline pkg/operator/vspherecontroller/vspherecontroller.go]: can we drop this commented code or are we leaving it in for historical reasons?

File: pkg/operator/vspherecontroller/vspherecontroller.go [hunk 2]
```diff
@@ -626,15 +647,20 @@ func (c *VSphereController) createCSISecret(
 		return fmt.Errorf("cloud config %s/%s does not contain key %q", cloudConfigNamespace, cloudConfig.Name, cloudConfig.Key)
 	}
 
-	var cfg vsphere.VSphereConfig
-	err = gcfg.ReadStringInto(&cfg, cfgString)
+	// Update this to also support YAML.
+	//var cfg vsphere.VSphereConfig
+	//err = gcfg.ReadStringInto(&cfg, cfgString)
```
Comment (rvanderp3) [inline pkg/operator/vspherecontroller/vspherecontroller.go]: can we drop this commented code or are we leaving it in for historical reasons?

File: pkg/operator/storageclasscontroller/storageclasscontroller.go [hunk 3]
```diff
@@ -91,24 +92,31 @@ func NewStorageClassController(
...
 		sc := resourceread.ReadStorageClassV1OrDie(c.manifest)
 		scState := c.scStateEvaluator.GetStorageClassState(sc.Provisioner)
 
-		policyName, syncResult := c.syncStoragePolicy(ctx, connection, apiDeps, scState)
-		if syncResult.CheckError != nil {
-			klog.Errorf("error syncing storage policy: %v", syncResult.Reason)
-			clusterCondition := "storage_class_sync_failed"
-			utils.InstallErrorMetric.WithLabelValues(string(syncResult.CheckStatus), clusterCondition).Set(1)
-			return syncResult, checks.ClusterCheckAllGood
+		// TODO: Iterate through each vcenter for storage policy.
```
Comment (rvanderp3) [inline pkg/operator/storageclasscontroller/storageclasscontroller.go]: does this todo need to be addressed?

File: pkg/operator/testlib/vcsim_framework.go [hunk 4]
```diff
@@ -83,6 +83,9 @@ func SetupSimulator(modelDir string) (*vclib.VSphereConnection, func(), error) {
 		Username: "admin@vsphere.local",
 		Password: "foobar",
 	}
+	/*connections := []*vclib.VSphereConnection{
+		conn,
+	}*/
```
Comment (rvanderp3) [inline pkg/operator/testlib/vcsim_framework.go]: can we drop this commented code or are we leaving it in for historical reasons?

File: pkg/operator/testlib/vcsim_framework.go [hunk 5]
```diff
@@ -155,7 +158,8 @@ folder = "/DC0/vm"
 [VirtualCenter "dc0"]
 datacenters = "DC0"
 `
-	err := gcfg.ReadStringInto(&cfg, data)
+	//err := gcfg.ReadStringInto(&cfg, data)
```
Comment (rvanderp3) [inline pkg/operator/testlib/vcsim_framework.go]: can we drop this commented code or are we leaving it in for historical reasons?

Comment (vr4manta) [reply pkg/operator/testlib/vcsim_framework.go]: this will be removed yes.  merge conflict resolving is fun.

File: pkg/operator/testlib/vcsim_framework.go (hunk 4 above)
Comment (vr4manta) [reply pkg/operator/testlib/vcsim_framework.go]: yes

File: pkg/operator/storageclasscontroller/storageclasscontroller.go (hunk 3 above)
Comment (vr4manta) [reply pkg/operator/storageclasscontroller/storageclasscontroller.go]: I believe I left this as a talking point with @gnufied when it gets reviewed.  I had some questions on "default" storage and such.  I think we really just want to map from infrastructure resource.  I'll leave this one for now and be sure to address before merge.

File: pkg/operator/vspherecontroller/vspherecontroller.go (hunk 2 above)
Comment (vr4manta) [reply pkg/operator/vspherecontroller/vspherecontroller.go]: These will be cleaned up shortly.  Still working on merging conflicts and fixing.

File: pkg/operator/vspherecontroller/vspherecontroller.go (hunk 1 above)
Comment (vr4manta) [reply pkg/operator/vspherecontroller/vspherecontroller.go]: These will be cleaned up shortly.  Still working on merging conflicts and fixing.

**Comment by vr4manta** [comment]
/hold
Fixing issues resulting from merge conflicts.

File: pkg/operator/storageclasscontroller/storageclasscontroller.go [hunk 6]
```diff
@@ -125,14 +133,15 @@ func (c *StorageClassController) Sync(ctx context.Context, connection *vclib.VSp
...
 		return "", checks.MakeClusterCheckResultPass()
 	}
 
 	// if we are running the checks after creating the policy successfully
 	// then lets run checks less frequently.
-	if !time.Now().After(c.nextCheck) && len(c.policyName) > 0 {
+	if !time.Now().After(c.nextCheck) && len(c.connPolicyNames[connection.Hostname]) > 0 {
 		klog.V(4).Infof("Returning without running any checks")
-		return c.policyName, checks.MakeClusterCheckResultPass()
+		return c.connPolicyNames[connection.Hostname], checks.MakeClusterCheckResultPass()
```
Comment (gnufied) [inline pkg/operator/storageclasscontroller/storageclasscontroller.go:144]: We will have to now probably create multiple storageclasses. One for each vcenter.

File: pkg/operator/storageclasscontroller/vmware.go [hunk 7]
```diff
@@ -76,12 +77,31 @@ func NewStoragePolicyAPI(ctx context.Context, connection *vclib.VSphereConnectio
...
+			}
+		}
+	} else {
+		if config != nil {
+			dcName = config.GetWorkspaceDatacenter()
+			dsName = config.GetDefaultDatastore()
+		} else {
+			return nil, fmt.Errorf("unable to determine default datastore from current config")
+		}
+	}
```
Comment (gnufied) [inline pkg/operator/storageclasscontroller/vmware.go:104]: So, this function is used only for populating migration datastore URL. Do we even need to read failure-domains for this one? Are workspace populated at all in multiple vcenter environment?

File: pkg/operator/storageclasscontroller/vmware.go [hunk 8]
```diff
@@ -135,12 +155,18 @@ func (v *storagePolicyAPI) createZonalStoragePolicy(ctx context.Context) (string
...
 
+	// Failure domains go across vCenters.  So really we want to only do failure domains that are for this connection's
+	// vCenter.
 	for _, failureDomain := range failureDomains {
-		dataCenter := failureDomain.Topology.Datacenter
-		dataStore := failureDomain.Topology.Datastore
-		err := v.attachTags(ctx, dataCenter, dataStore)
-		if err != nil {
-			aggregatedErrors = append(aggregatedErrors, err)
+		klog.V(4).Infof("Is failure domain %v part of this connection (%v): %v", failureDomain.Name, v.vcenterApiConnection.Hostname, failureDomain.Server == v.vcenterApiConnection.Hostname)
```
Comment (gnufied) [inline pkg/operator/storageclasscontroller/vmware.go:161]: Rather than doing this check here, we should ensure that we never run into a scenario that - this code is suuplied incorrect connection object.

File: assets/rbac/csi_driver_controller_role.yaml [hunk 9]
```diff
@@ -40,7 +40,7 @@ rules:
     resources: ["triggercsifullsyncs"]
     verbs: ["create", "get", "update", "watch", "list"]
   - apiGroups: ["cns.vmware.com"]
-    resources: ["cnsvspherevolumemigrations"]
+    resources: ["cnsvolumeinfoes"]
```
[the remark this answers is not in the document]
Comment (vr4manta) [reply assets/rbac/csi_driver_controller_role.yaml:43]: For some reason, the resource `cnsvspherevolumemigrations` was defined twice (See line 28).  I ran into an error were `cnsvolumeinfoes` was missing.

File: pkg/operator/storageclasscontroller/storageclasscontroller.go (hunk 6 above)
Comment (vr4manta) [reply pkg/operator/storageclasscontroller/storageclasscontroller.go:144]: Right.  The connPolicyNames is containing policy for each vCenter that we create.  Originally, first vCenter to create the policy set this older variable which made logic think second vCenter was already done.  So with this map, we are storing fact that each vCenter policy creation is complete.

File: pkg/operator/storageclasscontroller/vmware.go (hunk 7 above)
Comment (vr4manta) [reply pkg/operator/storageclasscontroller/vmware.go:104]: When migrating to multi vCenter support, I am having the cluster now use the YAML styled new vsphere cloud provider config.  In this config, there is no concept of datastores.  In the 4.12 and older OCP clusters, no FD is defined and has workspace datastore populated.  Are you ok with this only populating when single vCenter is in use?

Also note that with multi vCenter, the migration function in the vsphere-syncer will log:
`{"level":"error","time":"2024-07-23T19:38:24.181570104Z","caller":"syncer/util.go:366","msg":"volume-migration feature is not supported on Multi-vCenter deployment","TraceId":"b8986369-a457-4670-a848-adb11353edd5","stacktrace":"sigs.k8s.io/vsphere-csi-driver/v3/pkg/syncer.initVolumeMigrationService\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/pkg/syncer/util.go:366\nsigs.k8s.io/vsphere-csi-driver/v3/pkg/syncer.podAdded\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/pkg/syncer/metadatasyncer.go:1851\nsigs.k8s.io/vsphere-csi-driver/v3/pkg/syncer.InitMetadataSyncer.func7\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/pkg/syncer/metadatasyncer.go:527\nk8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/vendor/k8s.io/client-go/tools/cache/controller.go:232\nk8s.io/client-go/tools/cache.(*processorListener).run.func1\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/vendor/k8s.io/client-go/tools/cache/shared_informer.go:909\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:157\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:158\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:135\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:92\nk8s.io/client-go/tools/cache.(*processorListener).run\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/vendor/k8s.io/client-go/tools/cache/shared_informer.go:903\nk8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1\n\t/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:75"}`

Basically setting the migration url will most likely get ignored.

File: pkg/operator/storageclasscontroller/vmware.go (hunk 8 above)
Comment (vr4manta) [reply pkg/operator/storageclasscontroller/vmware.go:161]: Ok.  I can enhance the connection ctx to contain the FDs that are related to that connection.  I can populate this around connection creation time after we log into the connections.  Would that be a good approach?

File: pkg/operator/storageclasscontroller/storageclasscontroller.go [hunk 10]
```diff
@@ -91,24 +92,31 @@ func NewStorageClassController(
...
+			policyName, syncResult := c.syncStoragePolicy(ctx, connection, apiDeps, scState)
+			if syncResult.CheckError != nil {
+				klog.Errorf("error syncing storage policy: %v", syncResult.Reason)
+				clusterCondition := "storage_class_sync_failed"
+				utils.InstallErrorMetric.WithLabelValues(string(syncResult.CheckStatus), clusterCondition).Set(1)
+				return syncResult, checks.ClusterCheckAllGood
+			}
+			klog.V(2).Infof("Synced policy %v", policyName)
+			c.connPolicyNames[connection.Hostname] = policyName
+			c.policyName = policyName // This is the global name of policy.  may need to make it more logical to not set in loop.
```
[the remark this answers is not in the document]
Comment (vr4manta) [reply pkg/operator/storageclasscontroller/storageclasscontroller.go:118]: Great question.  This is where I was not sure.  I ended up creating a matching policy in each vCenter so that PVCs could be provisioned for image registry / monitoring.  We have one storage class (SC) defined but was under the assumption we had to have matching policy for each vCenter.

For limiting to one zone, lets discuss this on our next call.  I'd like to understand more about this requirement.

File: pkg/operator/vspherecontroller/checks/interface.go [hunk 11]
```diff
@@ -2,18 +2,22 @@ package checks
...
+	"github.com/openshift/library-go/pkg/operator/configobserver/featuregates"
 	check "github.com/openshift/vmware-vsphere-csi-driver-operator/pkg/operator/vclib"
 )
 
 type CheckArgs struct {
-	vmConnection *check.VSphereConnection
-	apiClient    KubeAPIInterface
+	vmConnection        []*check.VSphereConnection
+	apiClient           KubeAPIInterface
+	featureGates        featuregates.FeatureGate
```
Comment (gnufied) [inline pkg/operator/vspherecontroller/checks/interface.go:12]: Why are we supplying this featuregate when we aren't using it?

Comment (vr4manta) [reply pkg/operator/vspherecontroller/checks/interface.go:12]: We use it for a minimum of checking for multi vcenter: 
https://github.com/openshift/vmware-vsphere-csi-driver-operator/blob/5fc756c9695ccb96a5af5e869c42cad7bedef5c4/pkg/operator/vspherecontroller/vspherecontroller.go#L780
I wasn't sure if we want it to throw an error and prevent operator from functioning if the feature gate is not enabled.  I can also add other checks to verify fg set when we detect multi vcenter. If you have a preferred place, I can also put there.

Comment (gnufied) [reply pkg/operator/vspherecontroller/checks/interface.go:12]: But that is not the place where `CheckArgs` is being used from. My point is - this field in `CheckArgs` is never used (at least as far as I can tell).

Comment (vr4manta) [reply pkg/operator/vspherecontroller/checks/interface.go:12]: oh, i thought that guy extended the CheckArgs.  Let me double check.  I'll clean up if its been removed (or use when i refactor next week based on our talks).

File: pkg/operator/vclib/connection.go [hunk 12]
```diff
@@ -24,7 +30,76 @@ type VSphereConnection struct {
...
+	LegacyConfig *legacy.VSphereConfig
+}
+
+// LoadConfig load the desired config into this config object.
+func (c *VSphereConfig) LoadConfig(data string) error {
+	var err error
+	c.Config, err = vsphere.ReadConfig([]byte(data))
+	if err != nil {
+		return err
+	}
```
Comment (gnufied) [inline pkg/operator/vclib/connection.go:49]: I was just testing this and this code currently always throws:

```
I0726 20:00:55.719681       1 config.go:272] ReadConfig INI succeeded. INI-based cloud-config is deprecated and will be removed in 2.0. Please use YAML based cloud-config.
```

This will always be returned for upgraded clusters, since nothing is there to migrate these INI based configs to YAMLs. So this is one of the things we talked about in the call. Do we want to write a ticket or something for this?

Comment (gnufied) [reply pkg/operator/vclib/connection.go:49]: I am thinking we should rather check if we are dealing with YAML or ini and then load one or the other.

File: pkg/operator/vspherecontroller/vspherecontroller.go [hunk 13]
```diff
@@ -456,32 +462,35 @@ func (c *VSphereController) createVCenterConnection(ctx context.Context, infra *
...
-	username, ok := secret.Data[userKey]
-	if !ok {
-		return fmt.Errorf("error parsing secret %q: key %q not found", cloudCredSecretName, userKey)
-	}
-	passwordKey := cfg.Workspace.VCenterIP + "." + "password"
-	password, ok := secret.Data[passwordKey]
-	if !ok {
-		return fmt.Errorf("error parsing secret %q: key %q not found", cloudCredSecretName, passwordKey)
-	}
+	for _, vcenter := range infra.Spec.PlatformSpec.VSphere.VCenters {
```
Comment (gnufied) [inline pkg/operator/vspherecontroller/vspherecontroller.go:475]: Will - `infra.Spec.PlatformSpec.VSphere.VCenters` populated for clusters upgraded from older versions of OCP?

File: pkg/operator/vspherecontroller/vspherecontroller.go [hunk 14]
```diff
@@ -658,46 +669,61 @@ func (c *VSphereController) createCSISecret(
...
+			"${MIGRATION_DATASTORE_URL}": datastoreURL,
+			"${PASSWORD}":                password,
+			"${USER}":                    user,
+		} {
+			vcenterStr = strings.ReplaceAll(vcenterStr, pattern, value)
+		}
+		vcenters = vcenters + "\n" + vcenterStr
+	}
+
 	for pattern, value := range map[string]string{
```
Comment (gnufied) [inline pkg/operator/vspherecontroller/vspherecontroller.go:715]: So, in general we want secret's content to be "stable". Because - we do not want operator/controllers to be restarted randomly if order of vcenter names in config file changes. If I am not wrong - iterating a golang's map is essentially random and there is no guarantee order in which it will appear and as a result, I think secret can contain vcenters in any random order and this will cause unnecessary restart of controller.

File: pkg/operator/storageclasscontroller/vmware.go (hunk 8 above)
Comment (gnufied) [reply pkg/operator/storageclasscontroller/vmware.go:161]: I was thinking that - may be, we should first find all failure-domains which are on this vcenter and then iterate through them. That way, the intent of the code will be clearer to the reader that, we are trying to create storage policy on failure-domains associated with this vcenter.

File: pkg/operator/vspherecontroller/vspherecontroller.go [hunk 15]
```diff
@@ -626,23 +635,25 @@ func (c *VSphereController) createCSISecret(
...
+	err = cfg.LoadConfig(cfgString)
 	if err != nil {
-		return err
+		return fmt.Errorf("unable to fetch cloud provider config: %v", err)
 	}
 
-	storageApiClient := storageclasscontroller.NewStoragePolicyAPI(ctx, c.vSphereConnection, infra)
+	// Pass in first vcenter for now.  I think this logic is no longer valid, but need to confirm if we are wanting
+	// multi vcenter to support storage migration.
+	storageApiClient := storageclasscontroller.NewStoragePolicyAPI(ctx, c.vSphereConnections[0], infra)
```
Comment (gnufied) [inline pkg/operator/vspherecontroller/vspherecontroller.go:646]: So, I do not very much like just randomly picking first `vSphereConnection` as default one. This is what we were discussing - we will have to probably degrade clusters if more than one VC exists and migration needs to be supported. cc @rvanderp3

File: pkg/operator/storageclasscontroller/vmware.go (hunk 8 above)
Comment (vr4manta) [reply pkg/operator/storageclasscontroller/vmware.go:161]: Excellent.  I like that.  I'll mock that up.

File: pkg/operator/vclib/connection.go (hunk 12 above)
Comment (vr4manta) [reply pkg/operator/vclib/connection.go:49]: The library we are using will check if YAML and if that fails, it will go back to INI.  The logging for this library is excessive, but we also load this config a lot when we check.  It would be nice to limit how often we are generating this config object.

Another alternative is if multi vCenter support is disabled (looking at the featuregate which when GA will always be true), we can load the old style config (INI).  This way its at least limited to output the messages when we are looking for the correct config.

File: pkg/operator/storageclasscontroller/storageclasscontroller.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (gnufied) [reply pkg/operator/storageclasscontroller/storageclasscontroller.go:118]: Okay, so I think this is currently working as expected. We created a storage policy with same name in each failure-domain and it works. So discard my comment. I made this before trying volume provisioning.

File: pkg/operator/vspherecontroller/vspherecontroller.go (hunk 13 above)
Comment (vr4manta) [reply pkg/operator/vspherecontroller/vspherecontroller.go:475]: Good question.  I think this method is not pre-filtered to upgrade to FD infra.  I can add that change in to prevent backward compat issues.

**Comment by duanwei33** [comment]
Hi  @vr4manta, when performing the pre-merge testing, it said rebase needed when building the payload image, could you help check?
> pull request openshift/vmware-vsphere-csi-driver-operator#228 needs to be rebased to branch master

**Comment by vr4manta** [comment]
@duanwei33 sure.  I'll rebase after I finish fixing unit test.

File: pkg/operator/vspherecontroller/vspherecontroller.go (hunk 15 above)
Comment (vr4manta) [reply pkg/operator/vspherecontroller/vspherecontroller.go]: I've changed this to be related to each vCenter.  We talked about how this is used for migration and that migration will be disabled in multi vCenter.  Also, in the case of multiple failure domains in the same vCenter, the first FD is usually (or should be) the original FD so that we can make sure to migrate those old volumes.  It is also possible for this to only look at the old ini config to get the default settings.

File: pkg/operator/vspherecontroller/vspherecontroller.go (hunk 13 above)
Comment (vr4manta) [reply pkg/operator/vspherecontroller/vspherecontroller.go:473]: I have now pre-filtered the FD data.  The only down side I am thinking for these changes is where logic expecting no FDs to be defined will no longer get executed.

File: pkg/operator/vspherecontroller/checks/interface.go (hunk 11 above)
Comment (vr4manta) [reply pkg/operator/vspherecontroller/checks/interface.go]: I have cleaned this up.

File: pkg/operator/storageclasscontroller/vmware.go (hunk 8 above)
Comment (vr4manta) [reply pkg/operator/storageclasscontroller/vmware.go]: Ok, I have made changes for this.  Please feel free to review and let me know your thoughts.

File: pkg/operator/storageclasscontroller/vmware_test.go [hunk 16]
```diff
@@ -173,38 +198,73 @@ func TestZonalPolicyCreation(t *testing.T) {
 				t.Fatalf("error connecting to vcenter: %v", connError)
 			}
 
-			infra := tc.infra
+			for _, conn := range connections {
+				//TODO: I feel like we shouldn't create the storage policy by hand since we gotta create the FD info.
```
Comment (jcpowermac) [inline pkg/operator/storageclasscontroller/vmware_test.go:202]: Do we want to get rid of this todo? I guess it might not matter as much since its a test

Comment (vr4manta) [reply pkg/operator/storageclasscontroller/vmware_test.go:202]: Yes.  I'll clean this up.  Thanks for catching.

Comment (vr4manta) [reply pkg/operator/storageclasscontroller/vmware_test.go]: Cleaned up comments.

File: pkg/operator/vspherecontroller/vspherecontroller.go (hunk 14 above)
Comment (vr4manta) [reply pkg/operator/vspherecontroller/vspherecontroller.go:738]: I made the changes.  let me know your thoughts.

File: pkg/operator/vclib/connection.go [hunk 17]
```diff
@@ -37,13 +111,13 @@ var (
 	clientLock sync.Mutex
 )
```
Comment (gnufied) [inline pkg/operator/vclib/connection.go:113]: Shouldn't these 2 declarations be on top of the file?

Comment (vr4manta) [reply pkg/operator/vclib/connection.go:113]: yes.  I'll move these up.

File: pkg/operator/vspherecontroller/config.go [hunk 18]
```diff
@@ -19,7 +19,7 @@ type iniConfig struct {
 func newINIConfig(config string) (*iniConfig, error) {
 	// We ignore inline comments when reading the INI file
 	// because passwords may contain comment symbols (# and ;)
-	options := iniv1.LoadOptions{IgnoreInlineComment: true}
+	options := iniv1.LoadOptions{IgnoreInlineComment: true, PreserveSurroundedQuote: true}
```
Comment (gnufied) [inline pkg/operator/vspherecontroller/config.go:22]: Why did we had to use `PreserveSurroundedQuote` here? cc @bertinatto

Comment (vr4manta) [reply pkg/operator/vspherecontroller/config.go:22]: It was a change from the other PR.  there was logic that was forcing this after the fact and this fixes it i thought a little cleaner since I had to fix merge conflict: 
https://github.com/openshift/vmware-vsphere-csi-driver-operator/pull/239/files#diff-bd7b198fe282f7e27754079508639882be175cc903f563fb69f2c89fc6875a2aR698

File: pkg/operator/storageclasscontroller/vmware.go [hunk 19]
```diff
@@ -64,9 +65,10 @@ type storagePolicyAPI struct {
 
 var _ vCenterInterface = &storagePolicyAPI{}
 
-func NewStoragePolicyAPI(ctx context.Context, connection *vclib.VSphereConnection, infra *v1.Infrastructure) vCenterInterface {
+func NewStoragePolicyAPI(ctx context.Context, connection []*vclib.VSphereConnection, infra *v1.Infrastructure) vCenterInterface {
+	// TODO: Need to evaluate how to handle new storage policy in multi vcenter env
 	storagePolicyAPIClient := &storagePolicyAPI{
-		vcenterApiConnection: connection,
+		vcenterApiConnection: connection[0], // Hack until multi vcenter support is implemented
```
[the remark this answers is not in the document]
Comment (vr4manta) [reply pkg/operator/storageclasscontroller/vmware.go]: This has been removed.

**Comment by bertinatto** [comment]
I'd love to see the results of vSphere payload jobs before merging this. I'll schedule them here and add an hold to this PR. Feel free to remove the hold once we're confident with the results.

/hold

/payload-job periodic-ci-openshift-release-master-ci-4.18-upgrade-from-stable-4.17-e2e-vsphere-ovn-upgrade
/payload-job periodic-ci-openshift-release-master-nightly-4.18-e2e-vsphere-ovn
/payload-job periodic-ci-openshift-release-master-nightly-4.18-e2e-vsphere-ovn-csi
/payload-job periodic-ci-openshift-release-master-nightly-4.18-e2e-vsphere-ovn-serial
/payload-job periodic-ci-openshift-release-master-nightly-4.18-e2e-vsphere-ovn-techpreview
/payload-job periodic-ci-openshift-release-master-nightly-4.18-e2e-vsphere-ovn-techpreview-serial
/payload-job periodic-ci-openshift-release-master-nightly-4.18-e2e-vsphere-ovn-upi
/payload-job periodic-ci-openshift-release-master-nightly-4.18-e2e-vsphere-ovn-upi-serial
/payload-job periodic-ci-openshift-release-master-nightly-4.18-e2e-vsphere-static-ovn

**Comment by rvanderp3** [comment]
my apologies folks, i should have set the qe label.  me--

**Comment by jcpowermac** [comment]
/hold cancel

Not seeing anything in the payload test that gives me pause

**Comment by duanwei33** [comment]
See test step in https://issues.redhat.com/browse/STOR-1814
